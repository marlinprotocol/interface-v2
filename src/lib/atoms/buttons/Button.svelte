<script lang="ts">
	import type { ButtonModel } from '$lib/types/componentTypes';
	import { getButtonStyles } from '$lib/utils/helpers/componentHelper';
	import { getButtonSize } from '$lib/utils/helpers/componentHelper';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let variant: ButtonModel['variant'] = 'filled';
	export let size: ButtonModel['size'] = 'medium';
	export let styleClass = '';
	export let onclick: ButtonModel['onclick'] = undefined;
	export let disabled = false;
	export let loading = false;
	export let id = '';

	const buttonStyles = getButtonStyles(variant);
	const buttonSize = getButtonSize(size);
</script>

{#if variant.includes('icon')}
	<button
		{id}
		{disabled}
		on:click={onclick}
		class={cn('gap-1.5', buttonSize, buttonStyles, styleClass)}
	>
		{#if loading}
			<span data-testid="loading-spinner-icon" class="loading loading-spinner loading-sm" />
		{:else}
			<slot />
		{/if}
	</button>
{:else}
	<button
		{id}
		{disabled}
		on:click={onclick}
		class={cn('gap-1.5', buttonSize, buttonStyles, styleClass)}
	>
		{#if loading}
			<span data-testid="loading-spinner" class="loading loading-spinner loading-sm" />
		{/if}
		<slot />
	</button>
{/if}
