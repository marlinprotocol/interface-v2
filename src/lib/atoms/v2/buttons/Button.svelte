<script lang="ts">
	import type { ButtonModel } from '$lib/types/componentTypes';
	import { getButtonStyles } from '$lib/utils/v2/helpers/componentHelper';
	import { getButtonSize } from '$lib/utils/v2/helpers/componentHelper';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let variant: ButtonModel['variant'] = 'filled';
	export let size: ButtonModel['size'] = 'medium';
	export let styleClass = '';
	export let onclick: ButtonModel['onclick'] = undefined;
	export let disabled = false;
	export let loading = false;

	const buttonStyles = getButtonStyles(variant);
	const buttonSize = getButtonSize(size);
</script>

<button {disabled} on:click={onclick} class={cn('gap-1.5', buttonSize, buttonStyles, styleClass)}>
	{#if loading}
		<span data-testid="loading-spinner" class="loading loading-spinner loading-sm" />
	{/if}
	<slot />
</button>
