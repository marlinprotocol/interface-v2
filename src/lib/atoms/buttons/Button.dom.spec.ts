import Button from './Button.svelte';
import { describe, it, afterEach } from 'vitest';
import { cleanup, render, screen, fireEvent } from '@testing-library/svelte';
import html from '@playpilot/svelte-htm';

describe('Button', () => {
	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
	});

	it('renders a button successfully', () => {
		const { getByRole } = render(Button);
		expect(getByRole('button')).toMatchSnapshot();
	});

	it('renders a button with text passed as its child', async () => {
		const { findByText } = render(html`<${Button}>test</${Button}>`);

		expect((await findByText('test')).innerHTML.trim()).toBe('test');
	});

	it('renders a spinner when loading is true', async () => {
		const { rerender, getByTestId } = render(Button, {
			props: { loading: false }
		});

		expect(() => getByTestId('loading-spinner')).toThrow();
		await rerender({
			loading: true
		});
		expect(() => getByTestId('loading-spinner')).not.toThrow();
	});

	it('disables if the disabled prop is true', () => {
		const { getByRole } = render(Button, {
			props: { disabled: true }
		});

		expect(() => getByRole('button')).not.toThrow();
		expect(getByRole('button').hasAttribute('disabled')).toBe(true);
	});

	it('fires the onClick event when clicked', async () => {
		const onClickFunction = vi.fn();
		render(Button, {
			props: { onclick: onClickFunction }
		});

		const button = screen.getByRole('button');
		expect(onClickFunction).not.toHaveBeenCalled();
		await fireEvent.click(button);
		expect(onClickFunction).toHaveBeenCalled();
	});
});
