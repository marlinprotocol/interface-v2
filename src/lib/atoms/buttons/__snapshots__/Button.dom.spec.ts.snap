// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders a button successfully 1`] = `
<button
  class="gap-1.5 h-12 btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed !rounded-[100px] btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white"
>
   
</button>
`;
